File: Assignments/Python/Fundamentals/ChainingMethods/ChainingMethods.py

class User:
    bank_name = "First National Dojo"
    def __init__(self, name, email_address):
        self.name = name
        self.email = email_address
        self.account_balance = 500
    def make_deposit(self, amount):
        self.account_balance += amount
        return self
    def make_withdrawal(self, amount):
        self.account_balance -= amount
        return self
    def display_user_balance(self):
        print(f"{self.name} has {self.account_balance} dollars")
        return self
    def  transfer_money(self, other_user, amount):
        self.account_balance -= amount
        other_user.account_balance += amount
        return self



guido = User("Guido van Rossum", "guido@python.com")
monty = User("Monty Python", "monty@python.com")
liam = User("Liam Desmond", "CodingDojo@test.com")

guido.make_deposit(100).make_deposit(150).make_deposit(50).make_withdrawal(100).display_user_balance()
monty.make_deposit(400).make_deposit(100).make_withdrawal(200).make_withdrawal(50).display_user_balance()
liam.make_deposit(800).make_withdrawal(100).make_withdrawal(50).make_withdrawal(200).display_user_balance()
guido.transfer_money(liam, 200).display_user_balance()
liam.display_user_balance()

File: Assignments/MERN/OOP_and_Class_Components/Chaining_Methods/Chaining_Methods.js
class User {
    constructor(username, emailAddress) { 
        this.name = username
        this.email = emailAddress
        this.accountBalance = 0
    }
    makeDeposit(amount) {
        this.accountBalance += amount;
        return this;
    }
    makeWithdrawal(amount) {
        this.accountBalance -= amount;
        return this;
    }
    displayBalance() {
        console.log(this.accountBalance);
        return this;
    }
    transferMoney(otherUser, amount) {
        this.accountBalance -= amount;
        otherUser.accountBalance += amount;
        return this;
    }
}

const guido = new User("Guido van Rossum", "guido@python.com");
const monty = new User("Monty Python", "monty@python.com");
const karl = new User("Karl Pilkington, kpRadio@test.com")

guido.makeDeposit(100).makeDeposit(100).makeDeposit(50).makeWithdrawal(50).displayBalance()

monty.makeDeposit(100).makeDeposit(50).makeWithdrawal(10).makeWithdrawal(20).displayBalance()

karl.makeDeposit(500).makeWithdrawal(200).makeWithdrawal(50).makeWithdrawal(50).displayBalance()

guido.transferMoney(karl, 150).displayBalance()

karl.displayBalance()